src/sentry/snuba/metrics/datasource.py

```diff
@@ -692,9 +692,17 @@ def get_series(projects: Sequence[Project], query: QueryDefinition) -> dict:
         projects[0].organization_id, query, fields_in_entities, intervals, results
     )
 
+    result_groups = converter.translate_results()
+    # It can occur, when we make queries that are not ordered, that we end up with a number of
+    # groups that doesn't meet the limit of the query for each of the entities, and hence they
+    # don't go through the pruning logic resulting in a total number of groups that is greater
+    # than the limit, and hence we need to prune those excess groups
+    if len(result_groups) > query.limit.limit:
```

Comment:
nit: `query.limit` is now never None thanks to the `__post_init__` function setting it, but the type of `limit` is still `Optional[Limit]`, so the type checker of my IDE complains at this point. Maybe a combination of using `dataclasses.field(init=False)` and `dataclasses.InitVar` could fix that?

Reply:
I get your point and I agree,
I agree on `dataclasses.field(init=False)` not sure I understand why I would use `dataclasses.InitVar` in this case

Reply:
I was thinking of having an `InitVar` of type `Optional[Limit]` for the limit requested by the user, and then a `field(init=False)` of type `Limit` to store the actual limit computed by the logic in `__post_init__`.